src/algebra/module/linear_map.lean

```diff
@@ -210,6 +211,72 @@ def comp : M →ₗ[R] M₃ := ⟨f ∘ g, by simp, by simp⟩
 @[norm_cast]
 lemma comp_coe : (f : M₂ → M₃) ∘ (g : M → M₂) = f.comp g := rfl
 
+@[simp] theorem comp_id : f.comp id = f :=
+linear_map.ext $ λ x, rfl
+
+@[simp] theorem id_comp : id.comp f = f :=
+linear_map.ext $ λ x, rfl
+
+end
+
+section
+
+variables [semimodule R M] (f : M →ₗ[R] M)
+
+/-- Iteration of a linear map is a linear map -/
+def iterate : ℕ → (M →ₗ[R] M)
+| n := ⟨f^[n], by {
+    intros x y,
+    induction n with n ih,
+    { simp only [id.def, iterate_zero] },
+    { simp only [map_add, iterate_succ', comp_app, ih] },
+  }, by {
+    intros x y,
+    induction n with n ih,
+    { simp only [id.def, iterate_zero] },
+    { simp only [map_smul, iterate_succ', comp_app, ih] },
+  }⟩
+
+lemma iterate_succ (n) : (iterate f (n + 1)) = comp (iterate f n) f := rfl
+
+lemma iterate_zero : f.iterate 0 = id := rfl
+lemma iterate_one : f.iterate 1 = f :=
+by rw [iterate_succ, iterate_zero, id_comp]
+
+theorem iterate_add : ∀ (m n : ℕ), iterate f (m + n) = comp (iterate f m) (iterate f n)
+| m 0       := by rw [add_zero, iterate, iterate_zero, comp_id]
+| m (n + 1) := by { rw [← add_assoc, iterate_succ, iterate_add m n], refl }
+
+lemma iterate_succ' (n) : (iterate f (n + 1)) = comp f (iterate f n) :=
+by rw [add_comm, iterate_add, iterate_one]
+
+@[simp] lemma iterate_to_fun : ∀ n, (iterate f n).to_fun = nat.iterate f.to_fun n
```

Comment:
```suggestion
@[simp] lemma iterate_to_fun : ∀ n, iterate n f = (⇑f)^[n]
```

Reply:
This seems to have needed more changes; reverted until I can look more closely.